### PR TITLE
Move SEI80BU encryption out of the SEIBU_SOUND device and make it a d…

### PR DESCRIPTION
**Pull request by ajrhacker**
…evice of its own

Add MCFG_DEVICE_ROM to override region tag for device_rom_interface (nw)

### CODE REVIEW

**Comment by Tafoid** [comment]
Regression testing picked this up today:

airraid - airraid.cpp
3:           Fatal error: memory_bank::set_entry called with out-of-range entry 0
cshooter - airraid.cpp
3:           Fatal error: memory_bank::set_entry called with out-of-range entry 0

**Comment by Tafoid** [comment]
Also, I was going through some regression results for speed drops and noticed this one hit all drivers using this new SEI80BU Encrypted Z80 Interface device with a pretty substantial loss of performance (easily 20% up to 50% or so).  Notice the one set not using this new device (raidenk) is running essentially equal to previous day benchmark.  Is there anything that can be done to avoid the speed drop?

[seibu_results.txt](https://github.com/mamedev/mame/files/658586/seibu_results.txt)

**Comment by ajrhacker** [comment]
On Fri, Dec 16, 2016 at 8:38 PM Scott Stone <notifications@github.com>
wrote:

> Also, I was going through some regression results for speed drops and
> noticed this one hit all drivers using this new SEI80BU Encrypted Z80
> Interface device with a pretty substantial loss of performance (easily 20%
> up to 50% or so). Notice the one set not using this new device (raidenk) is
> running essentially equal to previous day benchmark. Is there anything that
> can be done to avoid the speed drop?
>

For a long time, MAME required encrypted opcodes to be pre-decrypted; this
was one of the main purposes of the DRIVER_INIT routine, and a lot of
drivers still use it for that. However, since Olivier Galibert implemented
AS_DECRYPTED_OPCODES in 2015, the devs have started to regard such
pre-decryption as deprecated, as evidenced by Vas Crabb's objections when I
tried to do that with bdream97. The now-preferred method of inserting
decryption handlers into the memory maps is, as has been proven in this
case, definitely less efficient because it disables the efficiency
shortcuts ordinarily provided to CPUs by direct_read_data (which is
coincidentally also used by the SEI80BU device).

In short, I see no easy way to regain efficiency here without offending
bigger names. OG has long been working on a major rewrite of the memory
system that supposedly will result in efficiency improvements across the
board, but I'm not sure when that will happen and how much impact it will
have in this particular case.

AJR

**Comment by awjackson** [comment]
As I understand it, the objection with bdream97 was that on that hardware the entire address space was encrypted, and predecrypting the ROMs didn't handle the (hypothetical?) case where the CPU fetched an opcode from RAM. In this case the encryption chip sits between the bus and one of the ROMs, and only access to that specific ROM goes through the decryption. Since the encryption definitely only affects ROM on this hardware, I think static decrypting at startup is perfectly acceptable here.

**Comment by ajrhacker** [comment]
On Sun, Dec 18, 2016 at 12:50 PM Alex W. Jackson <notifications@github.com>
wrote:

> In this case the encryption chip sits between the bus and one of the ROMs,
> and only access to that specific ROM goes through the decryption. Since the
> encryption definitely only affects ROM on this hardware, I think static
> decrypting at startup is perfectly acceptable here.
>

The encryption actually affects two ROM chips on Dynamite Duke and Raiden,
one of them being banked, so it isn't a simple matter of caching the
results. (device_rom_interface has a bankswitching callback, but that can
only be used when the entire space is banked.) I'm not inclined to take
your word about static decrypting remaining "perfectly acceptable" in cases
like this.

AJR
